debezium-embedded/src/main/java/io/debezium/embedded/AsyncEmbeddedEngine.java

```diff
@@ -0,0 +1,1266 @@
+                    clock,
+                    transformations);
+            task.initialize(taskContext); // Initialize Kafka Connect source task
+            tasks.add(new EngineSourceTask(task, taskContext)); // Create new DebeziumSourceTask
+        }
+    }
+
+    /**
+     * Starts the source tasks.
+     * The caller is responsible for handling possible error states.
+     * However, all the tasks are awaited to either start of fail.
+     *
+     * @param tasks {@link List<EngineSourceTask>} of tasks to be started
+     */
+    private void startSourceTasks(final List<EngineSourceTask> tasks) throws Exception {
+        LOGGER.debug("Starting source connector tasks.");
+        final ExecutorCompletionService<Void> taskCompletionService = new ExecutorCompletionService(taskService);
+        for (EngineSourceTask task : tasks) {
+            taskCompletionService.submit(() -> {
+                task.connectTask().start(task.context().config());
+                return null;
+            });
+        }
+
+        final long taskStartupTimeout = config.getLong(AsyncEngineConfig.TASK_MANAGEMENT_TIMEOUT_MS);
+        LOGGER.debug("Waiting max. for {} ms for individual source tasks to start.", taskStartupTimeout);
+        final int nTasks = tasks.size();
+        Exception error = null;
+        // To avoid leaked resources, we have to ensure that all tasks that were scheduled to start are really started before we continue with the execution in
+        // the main (engine) thread and change engine state. If any of the scheduled tasks has failed, catch the exception, wait for other tasks to start and then
+        // re-throw the exception and let engine stop already running tasks gracefully during the engine shutdown.
+        for (int i = 0; i < nTasks; i++) {
+            try {
+                final Future<Void> taskFuture = taskCompletionService.poll(taskStartupTimeout, TimeUnit.MILLISECONDS);
+                if (taskFuture != null) {
+                    taskFuture.get(); // we need to retrieve the results to propagate eventual exceptions
+                }
+                else {
+                    throw new InterruptedException("Time out while waiting for source task to start.");
+                }
+
+                LOGGER.debug("Started task #{} out of {} tasks.", i + 1, nTasks);
+            }
+            catch (Exception e) {
+                LOGGER.debug("Task #{} (out of {} tasks) failed to start. Failed with {}", i + 1, nTasks, e.getMessage());
+                if (LOGGER.isDebugEnabled()) {
+                    e.printStackTrace();
+                }
+
+                // Store only the first error.
+                if (error == null) {
+                    error = e;
+                }
+                continue;
+            }
+            LOGGER.debug("Calling connector callback after task is started.");
+            connectorCallback.ifPresent(DebeziumEngine.ConnectorCallback::taskStarted);
```

Comment:
Future improvement - the `taskStarted` should get info which task was started, id and config?